### DISCUSSION TITLE
Is $JS.ACK purposefully domain-unaware? Potential interference with leaf nodes

### DISCUSSION THREAD

**Discussion by cbecker** [nats-io/nats-server#6136] [Q&A] [answered]
I found this while debugging an issue that was caused by $JS.ACK not containing the domain it comes from (correct me if I'm wrong):

1. Assume we have 2 leaf nodes, each with its own domain `leaf_1` and `leaf_2`
2. Assume we have a single nats server, `server`,  where all leaf nodes connect
3. We create a stream `mystream` on `leaf_1` , and another on `leaf_2`. We want them to be independent.
4. For each stream, also create a durable consumer `leaf_{i}.mystream.myconsumer` 

The `$JS.{domain}.API` commands are correctly segregated by domain, which is great.

**But** the `$JS.ACK` commands are not, from what I see. 

--> The problem: ACK commands from a leaf node are sent back to other leaf nodes and confuse the consumers. 

Our workaround was to block the JS.ACK messages from/to the leaf nodes, but it seems a bit convoluted. Is there a plan to have ACK messages contain the domain they belong to? This will have some backward compatibility issues however.

Let me know if I got the flow right, I am still learning about NATS and its internals. Thanks!

**Comment by derekcollison** [comment, marked as answer]
You are correct, in the design on domains ACK and MSG gets (both direct and original) are not domain aware.

We plan to fix for sure.
